### PR TITLE
Avoid duplicating generators

### LINKED ISSUE #1980: Bad serialization ordering with conditionals
Given the test case:

```js

function bar() {
  return 123;
}

function fn(x) {
  if (!x) {
    return bar();
  }

  var foo = x.foo;
  if (!foo) {
    return 456;
  }
}

this.__optimize && __optimize(fn);

inspect = function() {
  return JSON.stringify(fn());
}
```

The output is

```js
var bar, fn;
(function () {
  var _$1 = this;

  var _$2 = _$1.Object;
  var _$3 = _$2.assign;

  var _1 = function (x) {
    var _$0 = _$3(x).foo; // <-- why is this defined here? it should be inside an if statement

    return !x ? 123 : !_$0 ? 456 : void 0;
  };

  var _C = function () {
    return 123;
  };

  var _0 = function () {
    return JSON.stringify(fn());
  };

  _$1.bar = _C;
  _$1.fn = _1;
  inspect = _0;
}).call(this);
```

If you remove the `if (!foo) {` from the original test case, it correctly serializes.

### PR DESCRIPTION
**Pull request by hermanventer**
Release note: none

Closes #1980 

Fixes a number of little oopsies: 
1) When composing two PNC instances, the priorEffects argument should only be composed into the fork of the second completion that is, or has, a normal completion.
2) The second completion's generator should not be appended to the generator of the saved effects of the first completion in composeWithSavedCompletion.
3) The generator of the effects that connects the normal branch of the first completion to the fork of the second completion should not be applied to the global effects, since if is captured by the composed completion.

### CODE REVIEW

File: src/methods/join.js [hunk 1]
```diff
@@ -164,29 +164,23 @@ export class JoinImplementation {
...
+          pnc.consequentEffects,
           c,
           !priorEffects ? newAlternateEffects : realm.composeEffects(priorEffects, newAlternateEffects),
           savedPathConditions,
           pnc.savedEffects
         );
       }
       invariant(pnc.alternate instanceof PossiblyNormalCompletion);
-      let new_alternate = this.composePossiblyNormalCompletions(realm, pnc.alternate, c);
+      let na = this.composePossiblyNormalCompletions(realm, pnc.alternate, c, priorEffects);
```
Comment (NTillmann) [inline src/methods/join.js]: What's the intuition behind the variable name "na"?

**Comment by trueadm** [comment]
Look good, I merged in master to see if another issue was fixed in our internal bundle and it seems to have. :)

**Comment by hermanventer** [comment]
These failures could be related to the problem that shows up with this repro:
`function fn(arg) {
  if (arg !== null) {
    if (arg.foo) {
      return 42;
    }
  }
}

if (global.__optimize) {
  __optimize(fn);
}

inspect = function() {
  return JSON.stringify([
      fn(null),
      fn({}),
      fn({foo: true}),
  ]);
};
`

**Comment by hermanventer** [comment]
This is now stale.

**Pull request closed without merging**